service/controller/v12/resource/instance/workingset.go

```diff
@@ -0,0 +1,45 @@
+package instance
+
+import "github.com/Azure/azure-sdk-for-go/services/compute/mgmt/2018-06-01/compute"
+
+type workingSet struct {
+	instanceToUpdate            *compute.VirtualMachineScaleSetVM
+	instanceToDrain             *compute.VirtualMachineScaleSetVM
+	instanceToReimage           *compute.VirtualMachineScaleSetVM
+	instanceAlreadyBeingUpdated *compute.VirtualMachineScaleSetVM
+}
```

Comment:
Proposal for slightly different API:

```go
func (ws *workingSet) WithInstanceToUpdate(instance *compute.VirtualMachineScaleSetVM) *workingSet {
	if ws == nil {
		ws = &workingSet{}
	}
	ws.instanceToUpdate = instance
	return ws
}

func (ws *workingSet) WithInstanceToDrain(instance *compute.VirtualMachineScaleSetVM) *workingSet {
	if ws == nil {
		ws = &workingSet{}
	}
	ws.instanceToDrain = instance
	return ws
}

func (ws *workingSet) WithInstanceToReimage(instance *compute.VirtualMachineScaleSetVM) *workingSet {
	if ws == nil {
		ws = &workingSet{}
	}
	ws.instanceToReimage = instance
	return ws
}

func (ws *workingSet) WithInstanceAlreadyBeingUpdated(instance *compute.VirtualMachineScaleSetVM) *workingSet {
	if ws == nil {
		ws = &workingSet{}
	}
	ws.instanceAlreadyBeingUpdated = instance
	return ws
}

func (ws *workingSet) IsWIP() bool {
	if ws == nil {
		return false
	}

	return ws.instanceToUpdate != nil ||
		ws.instanceToDrain != nil ||
		ws.instanceToReimage != nil ||
		ws.instanceAlreadyBeingUpdated != nil
}
```

This would then work as follows:
```go
func Test_WorkingSet(t *testing.T) {
	var ws *workingSet

	if ws.IsWIP() != false {
		t.Fatal("<nil>.IsWIP() == true")
	}

	ws = ws.WithInstanceToUpdate(&compute.VirtualMachineScaleSetVM{})

	if ws.IsWIP() != true {
		t.Fatal("workingSet{...}.IsWIP() == false")
	}

	ws = nil

	ws = ws.WithInstanceToDrain(&compute.VirtualMachineScaleSetVM{})

	if ws.IsWIP() != true {
		t.Fatal("workingSet{...}.IsWIP() == false")
	}

	ws = nil

	ws = ws.WithInstanceToReimage(&compute.VirtualMachineScaleSetVM{})

	if ws.IsWIP() != true {
		t.Fatal("workingSet{...}.IsWIP() == false")
	}

	ws = nil

	ws = ws.WithInstanceAlreadyBeingUpdated(&compute.VirtualMachineScaleSetVM{})

	if ws.IsWIP() != true {
		t.Fatal("workingSet{...}.IsWIP() == false")
	}

	ws = nil
}
```

Few pros with this:
* Readability as text / story is slightly nicer (opinionated of course, but still).
* You can combine different elements while still having one-line function to "set them" - given you declare variable once in the beginning.
* `IsWIP()` works with `nil` and value.
* Return in error cases can be usual `return nil, microerror.Mask(err)`

No strong opinions on this. Just a proposal. :slightly_smiling_face:

Reply:
I like it, applied